### PR TITLE
Add `optimisation_level` 3 using `GreedyPauliSimp`

### PR DESCRIPTION
**Pull request by sjdilkes**
Adds a new optimisation level "3" that uses `GreedyPauliSimp` targeting `zzphase` to optimise circuits. This calls `RemoveBarriers`, which is part of the motivation for adding it as a new level of optimisation. This mimics the changes to `pytket-qiskit`.

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 1]
```diff
@@ -726,6 +730,29 @@ def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
                     RemoveRedundancies(),
                 ]
             )
+        else:
+            passlist.extend(
+                [
+                    RemoveBarriers(),
+                    AutoRebase({OpType.CX, OpType.Rz, OpType.H}),
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:737]: I think we should allow more gates that `GreedyPauliSimp` can handle so the rebase can run faster. Also easier for the Pauli graph conversion.

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 2]
```diff
@@ -655,14 +657,16 @@ def rebase_pass(self) -> BasePass:
             allow_swaps=self.compilation_config.allow_implicit_swaps,
         )
 
-    def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
+    def default_compilation_pass(
+        self, optimisation_level: int = 2, timeout: int = 300
+    ) -> BasePass:
         """
         :param optimisation_level: Allows values of 0,1 or 2, with higher values
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:664]:
```suggestion
        :param optimisation_level: Allows values of 0,1,2 or 3, with higher values
```

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 3]
```diff
@@ -726,6 +730,29 @@ def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
...
+        else:
+            passlist.extend(
+                [
+                    RemoveBarriers(),
+                    AutoRebase({OpType.CX, OpType.Rz, OpType.H}),
+                    GreedyPauliSimp(
+                        allow_zzphase=True,
+                        only_reduce=True,
+                        thread_timeout=timeout,
+                        trials=10,
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:742]: When `GreedyPauliSimp` returns false, will the user know the cause, so they can, for example, try increasing the timeout?

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 4]
```diff
@@ -726,6 +730,29 @@ def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
                     RemoveRedundancies(),
                 ]
             )
+        else:
+            passlist.extend(
+                [
+                    RemoveBarriers(),
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:736]: Should we also issue a warning if the circuit contains barriers?

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 5]
```diff
@@ -655,14 +657,16 @@ def rebase_pass(self) -> BasePass:
             allow_swaps=self.compilation_config.allow_implicit_swaps,
         )
 
-    def default_compilation_pass(self, optimisation_level: int = 2) -> BasePass:
+    def default_compilation_pass(
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:660]: should also document the `timeout` parameter

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 6]
```diff
@@ -655,14 +657,16 @@ def rebase_pass(self) -> BasePass:
...
+        self, optimisation_level: int = 2, timeout: int = 300
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:661]: It seems the user cannot set a `timeout` when calling `get_compiled_circuit`. Any thoughts on this?

Comment (sjdilkes) [reply pytket/extensions/quantinuum/backends/quantinuum.py:661]: I've added `get_compiled_circuit` and `get_compiled_circuits` to `QuantinuumBackend` with a timeout parameter

File: pytket/extensions/quantinuum/backends/quantinuum.py (hunk 5 above)
Comment (sjdilkes) [reply pytket/extensions/quantinuum/backends/quantinuum.py:660]: done

File: pytket/extensions/quantinuum/backends/quantinuum.py (hunk 4 above)
Comment (sjdilkes) [reply pytket/extensions/quantinuum/backends/quantinuum.py:736]: good point, added to the new overloaded `get_compiled_circuit` method

File: pytket/extensions/quantinuum/backends/quantinuum.py (hunk 3 above)
Comment (sjdilkes) [reply pytket/extensions/quantinuum/backends/quantinuum.py:742]: They won't & I can't think of a good way of sharing this information in the code. Might just have to make it clear in the documentation?

File: pytket/extensions/quantinuum/backends/quantinuum.py (hunk 1 above)
Comment (sjdilkes) [reply pytket/extensions/quantinuum/backends/quantinuum.py:737]: done

File: pytket/extensions/quantinuum/backends/quantinuum.py (hunk 3 above)
Comment (yao-cqc) [reply pytket/extensions/quantinuum/backends/quantinuum.py:772]: I see. Can we print out some information in the C++ code? If you think that's possible, we should create an issue for this.

**Review by yao-cqc — APPROVED** [review]

**Pull request merged**